File: app/src/main/java/com/minibrowser/telemetry/TelemetryManager.java
package com.minibrowser.telemetry;

import android.content.Context;
import android.util.Log;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicLong;

public class TelemetryManager {
    private static final String TAG = "TelemetryManager";
    private static final ConcurrentHashMap<String, AtomicLong> counters = new ConcurrentHashMap<>();
    private static volatile boolean initialized = false;

    public static synchronized void init(Context context) {
        if (initialized) return;
        initialized = true;
        Log.i(TAG, "OpenTelemetry-lite Engine initialized for MiniBrowser");
    }

    public static void recordMetric(String metricName, long value) {
        counters.putIfAbsent(metricName, new AtomicLong(0));
        counters.get(metricName).addAndGet(value);
        Log.d(TAG, "Metric [" + metricName + "] updated: " + counters.get(metricName).get());
    }

    public static long getMetricValue(String metricName) {
        AtomicLong val = counters.get(metricName);
        return val != null ? val.get() : 0L;
    }

    public static void recordDownload() {
        recordMetric("downloads.completed", 1);
    }

    public static void recordAiQuery() {
        recordMetric("ai.queries.total", 1);
    }

    public static void recordBlockedDomain() {
        recordMetric("adblock.domains.blocked", 1);
    }

    public static void shutdown() {
        Log.i(TAG, "OpenTelemetry Engine shutdown completed");
        counters.clear();
        initialized = false;
    }
}

File: app/src/main/java/com/minibrowser/telemetry/TelemetryManager.kt
package com.minibrowser.telemetry

import android.content.Context
import android.util.Log
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicLong

object TelemetryManager {
    private const val TAG = "TelemetryManager"
    private val counters = ConcurrentHashMap<String, AtomicLong>()
    @Volatile
    private var initialized = false

    @JvmStatic
    fun init(context: Context) {
        if (initialized) return
        initialized = true
        Log.i(TAG, "OpenTelemetry-lite Engine initialized for MiniBrowser (Kotlin)")
    }

    @JvmStatic
    fun recordMetric(metricName: String, value: Long) {
        counters.putIfAbsent(metricName, AtomicLong(0))
        counters[metricName]?.addAndGet(value)
        Log.d(TAG, "Metric [$metricName] updated: ${counters[metricName]?.get()}")
    }

    @JvmStatic
    fun getMetricValue(metricName: String): Long {
        return counters[metricName]?.get() ?: 0L
    }

    @JvmStatic
    fun recordDownload() {
        recordMetric("downloads.completed", 1)
    }

    @JvmStatic
    fun recordAiQuery() {
        recordMetric("ai.queries.total", 1)
    }

    @JvmStatic
    fun recordBlockedDomain() {
        recordMetric("adblock.domains.blocked", 1)
    }

    @JvmStatic
    fun shutdown() {
        Log.i(TAG, "OpenTelemetry Engine shutdown completed")
        counters.clear()
        initialized = false
    }
}
